packages/bull-vault/src/AuctionBull.sol

```diff
@@ -0,0 +1,565 @@
+             * - if target WETH to have in euler less than current amount in euler, remove WETH from euler
+             * - deposit into crab, and pay auction traders wPowerPerp  
+             */
+            _executeCrabDeposit(_crabAmount, _wethTargetInEuler, _wethLimitPrice, ethInCrab);
+            {
+                uint256 remainingAmount = wPowerPerpAmount;
+                uint256 ordersLength = _orders.length;
+                for (uint256 i; i < ordersLength; ++i) {
+                    _transferToOrder(_orders[i], remainingAmount, _clearingPrice);
+                    if (remainingAmount > _orders[i].quantity) {
+                        remainingAmount = remainingAmount.sub(_orders[i].quantity);
+                    } else {
+                        break;
+                    }
+                }
+            }
+        } else {
+            IBullStrategy(bullStrategy).redeemCrabAndWithdrawWEth(_crabAmount, wPowerPerpAmount);
+
+            {
+                uint256 remainingAmount = wPowerPerpAmount;
+                uint256 ordersLength = _orders.length;
+                for (uint256 i; i < ordersLength; ++i) {
+                    _transferToOrder(_orders[i], remainingAmount, _clearingPrice);
+                    if (remainingAmount > _orders[i].quantity) {
+                        remainingAmount = remainingAmount.sub(_orders[i].quantity);
+                    } else {
+                        break;
+                    }
+                }
+            }
+
+            _rebalanceLeverageComponent(_wethTargetInEuler, _wethLimitPrice);
+        }
+    }
+
+    function _executeCrabDeposit(uint256 _crabAmount, uint256 _wethTargetInEuler, uint256 _wethLimitPrice, uint256 _ethInCrab) internal {
+        uint256 wethFromEuler;
+        uint256 ethNeededForCrab = _crabAmount.wdiv(IERC20(crab).totalSupply()).wmul(_ethInCrab);
+        uint256 wethInCollateral = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
+        if (_wethTargetInEuler > wethInCollateral) {
+            uint256 wethToBuy;
+            uint256 usdcToBorrow;
+            {
+                uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+                wethToBuy = _wethTargetInEuler.sub(wethInCollateral);
+                usdcToBorrow = wethToBuy.wmul(ethUsdPrice).div(1e12);
+            }
+            _exactOutFlashSwap(
+                usdc,
+                weth,
+                3000,
+                wethToBuy,
+                usdcToBorrow.wmul(_wethLimitPrice),
+                uint8(FLASH_SOURCE.FULL_REBALANCE_BORROW_USDC_BUY_WETH),
+                abi.encodePacked(wethToBuy, usdcToBorrow)
+            );
+        } else {
+            wethFromEuler = wethInCollateral.sub(_wethTargetInEuler);
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(0, wethFromEuler);
```

Comment:
we need to figure out the difference between wethFromEuler and the ethNeededForCrab as well as wethFromAuction to figure out how much need to borrow additionally